### ISSUE TITLE
New sbx file processing trouble

### ISSUE THREAD

**Issue by ronan-dasilva** [MouseLand/suite2p#287]
Hey!

Really cool to see that suite2p now has easy support to different file formats in the GUI!

I am having some trouble with running suite2p on sbx files, however. For some reason suite2p considers 724 columns of my videos as "dead columns" and removes them from the processing, leaving in fact just one of the two deadbands we usually get from Scanbox. Any idea why thus happens, or where could I look to solve this?

Thanks!

**Comment by carsen-stringer** [comment]
@jcouto wrote the scanbox extension and auto "dead" column removal and might be able to help

**Comment by jcouto** [comment]
Hi,
did you record with unidirectional scanning? The default for the GUI right now is to try to discover the number of columns that are incomplete (but that may fail for unidirectional scanning).

I'll send a pull request that reads the scan mode from the scanbox mat file so this does not happen in unidirectional scanning. In the meantime you can still run suite2p by setting the sbx_ndeadcols option to zero or using [this](https://github.com/MouseLand/suite2p/blob/master/jupyter/run_pipeline_sbx.ipynb) example jupyter notebook. Please let me know if that works for you.

**Comment by ronan-dasilva** [comment]
Yes, it seems that we are using unidirectional scanning by default.
I'll try setting sbx_ndeadcols to zero as you suggested and post what happens.

Thanks!

**Comment by ronan-dasilva** [comment]
So, I tried running the code from the jupyter nb, but I am getting the following error:

`Traceback (most recent call last):

  File "C:\Users\Ronan\Desktop\untitled0.py", line 55, in <module>
    opsEnd = run_s2p.run_s2p(ops=ops,db={})

  File "C:\Users\Ronan\Miniconda3\envs\suite2p\lib\site-packages\suite2p\run_s2p.py", line 204, in run_s2p
    ops1 = sbx.sbx_to_binary(ops)

  File "C:\Users\Ronan\Miniconda3\envs\suite2p\lib\site-packages\suite2p\io\sbx.py", line 99, in sbx_to_binary
    ops1, sbxlist, reg_file, reg_file_chan2 = utils.find_files_open_binaries(ops1)

  File "C:\Users\Ronan\Miniconda3\envs\suite2p\lib\site-packages\suite2p\io\utils.py", line 211, in find_files_open_binaries
    fs, ops2 = get_sbx_list(ops1[0])

  File "C:\Users\Ronan\Miniconda3\envs\suite2p\lib\site-packages\suite2p\io\utils.py", line 48, in get_sbx_list
    look_one_level_down = ops['look_one_level_down'])

  File "C:\Users\Ronan\Miniconda3\envs\suite2p\lib\site-packages\suite2p\io\utils.py", line 24, in search_for_ext
    raise OSError('Could not find files, check path [{0}]'.format(rootdir))

OSError: Could not find files, check path [D]`

I am not sure why this happens, since when I use `search_for_ext(ops['data_path']) I am perfectly able to find the file and extract all the info as per the example code...

Help?

**Comment by jcouto** [comment]
> OSError: Could not find files, check path [**D**]

Probably ops['data_path'] is a string instead of a list.  

Each folder should be wrapped in square brackets: 
`ops['data_path'] = ['D:\\PATH']`

**Comment by ronan-dasilva** [comment]
Ok, so that helped!

Now I am running into an error further down on the processing, sometimes during ROI detection and extraction, sometimes during spike deconvolution. I am not quite sure what is causing this, but I keep getting a memory error that keeps repeating itself until I interrupt the process. Here's a piece of the error message:

'''python
FOUND OPS IN D:/Ronan/Imaging_Scope_1/RdS_031/img10k/RL/part4/a01\suite2p\plane0
FOUND BINARIES: c:/suite2p\suite2p\plane0\data.bin
SKIPPING REGISTRATION FOR ALL PLANES...
>>>>>>>>>>>>>>>>>>>>> PLANE 0 <<<<<<<<<<<<<<<<<<<<<<
----------- ROI DETECTION AND EXTRACTION
Binning movie in chunks of length 14
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "C:\Users\Ronan\Miniconda3\envs\suite2p\lib\multiprocessing\spawn.py", line 105, in spawn_main
    exitcode = _main(fd)
  File "C:\Users\Ronan\Miniconda3\envs\suite2p\lib\multiprocessing\spawn.py", line 114, in _main
    prepare(preparation_data)
  File "C:\Users\Ronan\Miniconda3\envs\suite2p\lib\multiprocessing\spawn.py", line 225, in prepare
    _fixup_main_from_path(data['init_main_from_path'])
  File "C:\Users\Ronan\Miniconda3\envs\suite2p\lib\multiprocessing\spawn.py", line 277, in _fixup_main_from_path
    run_name="__mp_main__")
  File "C:\Users\Ronan\Miniconda3\envs\suite2p\lib\runpy.py", line 263, in run_path
    pkg_name=pkg_name, script_name=fname)
  File "C:\Users\Ronan\Miniconda3\envs\suite2p\lib\runpy.py", line 96, in _run_module_code
    mod_name, mod_spec, pkg_name, script_name)
  File "C:\Users\Ronan\Miniconda3\envs\suite2p\lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "C:\Users\Ronan\desktop\sbx_s2p.py", line 59, in <module>
    opsEnd = run_s2p.run_s2p(ops=ops,db={})
  File "C:\Users\Ronan\Miniconda3\envs\suite2p\lib\site-packages\suite2p\run_s2p.py", line 282, in run_s2p
    ops1[ipl] = extract.detect_and_extract(ops1[ipl])
  File "C:\Users\Ronan\Miniconda3\envs\suite2p\lib\site-packages\suite2p\extraction\extract.py", line 84, in detect_and_extract
    ops, stat = sparsedetect.sparsery(ops)
  File "C:\Users\Ronan\Miniconda3\envs\suite2p\lib\site-packages\suite2p\detection\sparsedetect.py", line 337, in sparsery
    rez, max_proj = get_mov(ops)
  File "C:\Users\Ronan\Miniconda3\envs\suite2p\lib\site-packages\suite2p\detection\sparsedetect.py", line 39, in get_mov
    mov = np.zeros((ops['nbinned'], Lyc, Lxc), np.float32)
MemoryError
{}
FOUND OPS IN D:/Ronan/Imaging_Scope_1/RdS_031/img10k/RL/part4/a01\suite2p\plane0
FOUND BINARIES: c:/suite2p\suite2p\plane0\data.bin
SKIPPING REGISTRATION FOR ALL PLANES...
>>>>>>>>>>>>>>>>>>>>> PLANE 0 <<<<<<<<<<<<<<<<<<<<<<
'''

I tried running this in two different files and got more or less the same type of error... This never happened when I would try running through the GUI :(

**Comment by ronan-dasilva** [comment]
I tried running the script one more time to catch the beginning of the infinite loop. This time it ran up to the spike deconvolution and then it just restarted `run_s2p()` several times:

```python
(suite2p) C:\Users\Ronan>python desktop/sbx_s2p.py
{}
FOUND OPS IN D:/Ronan/Imaging_Scope_1/RdS_031/img10k/RL/part4/a01\suite2p\plane0
FOUND BINARIES: c:/suite2p\suite2p\plane0\data.bin
SKIPPING REGISTRATION FOR ALL PLANES...
 PLANE 0 <<<<<<<<<<<<<<<<<<<<<<
----------- ROI DETECTION AND EXTRACTION
Binning movie in chunks of length 14
Binned movie [5199,765,724], 142.05 sec.
NOTE: estimated spatial scale ~12 pixels, time epochs 4.33, threshold 43.32
0 ROIs, score=272.21
Found 206 ROIs, 422.25 sec
NOTE: applying classifier C:\Users\Ronan\Miniconda3\envs\suite2p\lib\site-packages\suite2p\extraction\../classifiers/classifier_user.npy
C:\Users\Ronan\Miniconda3\envs\suite2p\lib\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
After removing overlaps, 184 ROIs remain
Masks made in 5.43 sec.
Extracted fluorescence from 184 ROIs in 72800 frames, 260.03 sec.
----------- Total 704.05 sec.
----------- SPIKE DECONVOLUTION
{}
{}
{}
{}
{}
{}
FOUND OPS IN D:/Ronan/Imaging_Scope_1/RdS_031/img10k/RL/part4/a01\suite2p\plane0
FOUND BINARIES: c:/suite2p\suite2p\plane0\data.bin
SKIPPING REGISTRATION FOR ALL PLANES...
 PLANE 0 <<<<<<<<<<<<<<<<<<<<<<
----------- ROI DETECTION AND EXTRACTION
Binning movie in chunks of length 14
FOUND OPS IN D:/Ronan/Imaging_Scope_1/RdS_031/img10k/RL/part4/a01\suite2p\plane0
FOUND BINARIES: c:/suite2p\suite2p\plane0\data.bin
SKIPPING REGISTRATION FOR ALL PLANES...
FOUND OPS IN D:/Ronan/Imaging_Scope_1/RdS_031/img10k/RL/part4/a01\suite2p\plane0
 PLANE 0 <<<<<<<<<<<<<<<<<<<<<<
----------- ROI DETECTION AND EXTRACTION
FOUND BINARIES: c:/suite2p\suite2p\plane0\data.bin
SKIPPING REGISTRATION FOR ALL PLANES...
 PLANE 0 <<<<<<<<<<<<<<<<<<<<<<
----------- ROI DETECTION AND EXTRACTION
Binning movie in chunks of length 14
Binning movie in chunks of length 14
FOUND OPS IN D:/Ronan/Imaging_Scope_1/RdS_031/img10k/RL/part4/a01\suite2p\plane0
FOUND BINARIES: c:/suite2p\suite2p\plane0\data.bin
SKIPPING REGISTRATION FOR ALL PLANES...
 PLANE 0 <<<<<<<<<<<<<<<<<<<<<<
----------- ROI DETECTION AND EXTRACTION
FOUND OPS IN D:/Ronan/Imaging_Scope_1/RdS_031/img10k/RL/part4/a01\suite2p\plane0
Binning movie in chunks of length 14
FOUND BINARIES: c:/suite2p\suite2p\plane0\data.bin
SKIPPING REGISTRATION FOR ALL PLANES...
FOUND OPS IN D:/Ronan/Imaging_Scope_1/RdS_031/img10k/RL/part4/a01\suite2p\plane0
 PLANE 0 <<<<<<<<<<<<<<<<<<<<<<
----------- ROI DETECTION AND EXTRACTION
FOUND BINARIES: c:/suite2p\suite2p\plane0\data.bin
SKIPPING REGISTRATION FOR ALL PLANES...
 PLANE 0 <<<<<<<<<<<<<<<<<<<<<<
----------- ROI DETECTION AND EXTRACTION
Binning movie in chunks of length 14
Binning movie in chunks of length 14
```

Just in case, here's the complete script I am trying to run:

```python
from suite2p import run_s2p
from suite2p.io.utils import search_for_ext
from suite2p.io.sbx import sbx_get_info

datapath = [['D:/Ronan/Imaging_Scope_1/RdS_031/img10k/RL/part4/a01']]

for data in datapath:
    ops = run_s2p.default_ops()
    ops['data_path'] = data
    
    fname = search_for_ext(ops['data_path'][0],'sbx')
    info = sbx_get_info(fname[0])
    
    #number of planes
    nplanes = 1
    if info.volscan and not isinstance(info.otwave,int):
        nplanes = len(info.otwave)
    
    #number of channels
    nchannels = 2 if info.channels == 1 else 1
    
    #aspect ratio and mag
    magidx = info.config.magnification -1
    mag = info.config.magnification_list[magidx]
    scale_x = info.calibration[magidx].x
    scale_y = info.calibration[magidx].y
    
    factor = 2 if info.scanmode == 0 else 1
    fs = factor*info.resfreq/info.config.lines/nplanes
    ops['input_format'] = 'sbx'
    if info.scanmode == 1:
        #do not remove dead columns
        ops['sbx_ndeadcols'] = 0
    else:
        #remove dead columns when in bidi mode
        ops['sbx_ndeadcols'] = 1
        
    #set suite2p ops
    ops['fs'] = fs/nplanes
    ops['nplanes'] = nplanes
    ops['nchannels'] = nchannels
    ops['aspect'] = scale_x/scale_y
    ops['tau'] = 0.7
    ops['sparse_mode'] = 1
    ops['fast_disk'] = 'c:/suite2p'
    #extra info
    ops['magnification'] = mag
    ops['scale_x'] = scale_x
    ops['scale_y'] = scale_y
    
    #run exp
    opsEnd = run_s2p.run_s2p(ops=ops,db={})
```

**Comment by jcouto** [comment]
Hi Ronan, 

did you get this to work? You may try to delete the c:/suite2p\suite2p or previously saved folders to see if that solves it.  

If it still does not work I am happy to give it a go if you send me a link with the dataset.

Best, Joao

**Comment by ronan-dasilva** [comment]
Hi João!

No, I haven't yet been able to get it running. I tried removing previously saved folders, but that didn't help. 

I have also tried converting my sbx to h5 files and run s2p from code, but I couldn't get that to work properly either, even though running it from the GUI works just fine... 
I am really not sure what is going on :( 
I could definitely send you a link to the dataset, but it is quite heavy (85Gb). Would that be OK? I could send you a link be email.

Thanks for the help!

Ronan

**Comment by ronan-dasilva** [comment]
Hi!

So, I finally found a work around. For some reason, when I would try to process by the GUI it wouldn't have any problem running the process until the end, but whenever I tried to to use `run_s2p.run_s2p()` I would get into this multiple-repeats problem.

So I just did exactly what the GUI does when it runs the process: instead of calling `run_s2p.run_s2p(ops=ops,db=db)` I used `os.system('python -u -W ignore -m suite2p --ops ops.npy --db db.npy')`.

And by doing so, the process run smoothly until the end and I got my pretty ROIs, fluorescence and spikes.

It really bugs me that I have to use this work around, but I guess that will do for the moment... haha

Thanks for all the help!
